phoenix-core-server/src/main/java/org/apache/phoenix/replication/ReplicationLogDiscovery.java

```diff
@@ -0,0 +1,501 @@
+        LOG.info("Starting new files processing for round: {}", replicationRound);
+        long startTime = EnvironmentEdgeManager.currentTime();
+        List<Path> files = replicationLogTracker.getNewFilesForRound(replicationRound);
+        LOG.info("Number of new files for round {} is {}", replicationRound, files.size());
+        while (!files.isEmpty()) {
+            processOneRandomFile(files);
+            files = replicationLogTracker.getNewFilesForRound(replicationRound);
+        }
+        long duration = EnvironmentEdgeManager.currentTime() - startTime;
+        LOG.info("Finished new files processing for round: {} in {}ms", replicationRound, duration);
+        getMetrics().updateTimeToProcessNewFiles(duration);
+    }
+
+    /**
+     * Processes all files (older than 1 round time) in the in-progress directory.
+     * Continuously processes files until no in-progress files remain.
+     * @throws IOException if there's an error during file processing
+     */
+    protected void processInProgressDirectory() throws IOException {
+        // Increase the count for number of times in progress directory is processed
+        getMetrics().incrementNumInProgressDirectoryProcessed();
+        LOG.info("Starting in progress directory processing");
+        long startTime = EnvironmentEdgeManager.currentTime();
+        long oldestTimestampToProcess = replicationLogTracker.getReplicationShardDirectoryManager()
+                .getNearestRoundStartTimestamp(EnvironmentEdgeManager.currentTime())
+                - getReplayIntervalSeconds() * 1000L;
+        List<Path> files = replicationLogTracker.getOlderInProgressFiles(oldestTimestampToProcess);
+        LOG.info("Number of In Progress files with oldestTimestampToProcess {} is {}",
+                oldestTimestampToProcess, files.size());
+        while (!files.isEmpty()) {
+            processOneRandomFile(files);
+            files = replicationLogTracker.getOlderInProgressFiles(oldestTimestampToProcess);
+        }
+        long duration = EnvironmentEdgeManager.currentTime() - startTime;
+        LOG.info("Finished in-progress files processing in {}ms", duration);
+        getMetrics().updateTimeToProcessInProgressFiles(duration);
+    }
+
+    /**
+     * Processes a single random file from the provided list.
+     * Marks the file as in-progress, processes it, and marks it as completed or failed.
+     * @param files - List of files from which to select and process one randomly
+     */
+    private void processOneRandomFile(final List<Path> files) throws IOException {
+        // Pick a random file and process it
+        Path file = files.get(ThreadLocalRandom.current().nextInt(files.size()));
+        Optional<Path> optionalInProgressFilePath = Optional.empty();
+        try {
+            optionalInProgressFilePath = replicationLogTracker.markInProgress(file);
+            if (optionalInProgressFilePath.isPresent()) {
+                processFile(file);
```

Comment:
I think we need to pass `optionalInProgressFilePath.get()` here since markInProgress can rename the file